### PR TITLE
RUMM-541 Fix: URLSessionSwizzler in iOS11/12

### PR DESCRIPTION
**Pull request by buranmert**
### What and why?

We swizzle `dataTaskWithURL` and `dataTaskWithRequest` as part of auto-instrumentation. 
iOS 11/12 calls `dataTaskWithRequest` FROM `dataTaskWithURL`, that leads to duplicate execution of our swizzled implementation.

This case was not handled ❌ 

### How?

This PR contains 
1. the fix for the problem
2. also some minor refactoring based on https://github.com/DataDog/dd-sdk-ios/pull/139

#### Fix

We didn't need any change in `dataTaskWithRequest` implementation.
Now, in `dataTaskWithURL`, after creating the `task` object we check if `task.originalRequest` has interception headers.
If it has then that means **the task is already intercepted and we should not intercept it once again.**

```swift
/// dataTaskWithURL
let task = currentTypedImp(impSelf, Self.selector, impURL, modifiedCompletion)
if var request = task.originalRequest,
    request.add(newHTTPHeaders: interceptionResult.httpHeaders) {
    /// request.add(...) succeeded -> interception needed
    swizzleResumeAndAddPayload()
} else {
    /// do NOT intercept!
    doNothing()
}
return task
```

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

### CODE REVIEW

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift [hunk 1]
```diff
@@ -156,26 +163,33 @@ internal class URLSessionSwizzler {
...
         for pair in httpHeaders {
             if modifiedRequest.value(forHTTPHeaderField: pair.key) == nil {
                 modifiedRequest.setValue(pair.value, forHTTPHeaderField: pair.key)
+            } else {
+                return false
             }
         }
-        return modifiedRequest
+        self = modifiedRequest
+        return true
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:176]: it doesn't change `self`

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift [hunk 2]
```diff
@@ -11,39 +11,35 @@ import Foundation
...
+    case completed(URLResponse?, Error?)
 }
 
 internal class URLSessionSwizzler {
     let dataTaskWithURL: DataTaskWithURL
     let dataTaskwithRequest: DataTaskWithRequest
-    let resume: Resume
+    /// resume must be shared among URLSessionSwizzler instances
+    /// to avoid swizzling task.resume() multiple times
+    static let resume = Resume()
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:33]: we took that decision verbally but the code allows multiple `URLSessionSwizzler` instances. 
either we can 
1. hope that the next developer will be aware of our verbal discussion 
2. or we can write code to prevent them from mistakes

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift [hunk 3]
```diff
@@ -156,26 +163,33 @@ internal class URLSessionSwizzler {
...
         }
-        return modifiedRequest
+        self = modifiedRequest
+        return true
     }
 }
 
-/// payload is a TaskObserver, executed in task.resume() and completion
+/// payloads is an array TaskObservers, each one is executed in task.resume() and completion
 private extension URLSessionTask {
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:181]: because, as we discussed Friday, we will swizzle twice (once for tracing, once for RUM)

File: Tests/DatadogTests/Datadog/Core/AutoInstrumentation/URLSessionSwizzlerTests.swift [hunk 4]
```diff
@@ -103,54 +139,46 @@ class URLSessionSwizzlerTests: XCTestCase {
...
-            return (taskObserver: observer, httpHeaders: self.modifiedHTTPHeaders)
-        }
-
-        swizzler.swizzleOnce(using: interceptor)
-
-        session.dataTask(with: mockURLRequest) { _, _, _ in }
-    }
-
-    func test_dataTask_urlCompletion_neverIntercept() {
+    func test_dataTask_requestCompletion_alreadyTracedRequest() {
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Tests/DatadogTests/Datadog/Core/AutoInstrumentation/URLSessionSwizzlerTests.swift:142]: yes, this is correct
i'm changing this part

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:33]: or

3. We can enforce it as part of the architecture that there's no reason to swizzle twice (i.e. by accepting `URLSessionSwizzler` instance in `TracingAutoInstrumentation.init()` and `RUMAutoInstrumentation.init()`).

which IMO would be a good and pragmatic trade-off between the code complexity and safety. Otherwise, we'll end up with self-defending unit tests and eventually 2 sort of error messages printed out to the user:
* _"XXX auto-instrumentation will not work"_,
* _"XXX auto-instrumentation will partially work"_.

IMO no reason to complicate this stuff.

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:176]: 👍

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:33]: as we discussed before, we will have one `URLSessionSwizzler` but we will run `swizzler.swizzle(...)` twice.
that way we can make `TracingAutoInstrumentation` and `RUMAutoInstrumentation` separate and we won't need a new type such as `CombinedInterceptor` as Android does.

this simplifies the code

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:181]: We agreed on [one swizzler instance](https://github.com/DataDog/dd-sdk-ios/pull/149#discussion_r443440028), but OK, this is another leak enforced by the hooks-approach: now the `URLSessionSwizzler` needs to know there may be multiple interceptors. And this needs to be tested in `URLSessionSwizzlerTests`.

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:181]: you are confused with swizzling and swizzlers.
can you please explain what you mean by hooks-approach exactly?

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift [hunk 5]
```diff
@@ -57,28 +53,39 @@ internal class URLSessionSwizzler {
...
                     }
+                    /// NOTE: RUMM-489 in iOS 11/12 dataTaskWithURL: calls dataTaskWithRequest: internally
+                    /// we need to check if the originalRequest already has interceptor headers
+                    /// if so, we don't intercept this request
                     let task = currentTypedImp(impSelf, Self.selector, impURL, modifiedCompletion)
-                    try? resumeSwizzler.swizzleIfNeeded(in: task)
-                    task.payload = interceptionResult.taskObserver
-                    blockTask = task
+                    if var request = task.originalRequest,
+                        request.add(newHTTPHeaders: interceptionResult.httpHeaders) {
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:81]: this is addressed in the comment below

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:33]: Less code doesn't mean simpler code. It's definitely less complex if concerns are separated → combining interceptors as a distinct unit of work. It would be also healthier for unit tests as they won't be mixing everything up like now: interception logic + interceptors iteration in single test.

But at the end, it's the same business for now.

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:181]: > can you please explain what you mean by hooks-approach exactly?

The [one](https://github.com/DataDog/dd-sdk-ios/blob/buranmert/RUMM-489-urlsession-hooks/Sources/Datadog/Tracing/TracingAutoInstrumentation.swift) based on passing enum-scoped static functions down to the `URLSession` execution stack without a clear ownership:
```swift
internal enum TracingRequestInterceptor {
    static func build(with tracedHosts: Set<URL>) -> RequestInterceptor

    static func tracingTaskObserver(
        tracer: DDTracer, 
        spanContext: DDSpanContext, 
        operationName: String
    ) -> TaskObserver
}
```
It comes from a branch named `-urlsession-hooks`.

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:181]: i think i see your point.
i'm not saying the current approach is perfect, but the idea is writing and testing it once and using it everywhere.

if we follow [this way](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-452-propose-swizzling-refactoring/Tests/DatadogTests/Datadog/Core/Autoinstrumentation/URLSessionSwizzlerTests.swift) then we will need to test twice if `TracingInterceptor` and `RUMInterceptor` keep track of intercepted tasks properly.
here we implement and test this within `URLSessionSwizzler` once so that interceptors won't need to know `URLSession` internals.

IOW, we can have 1 dumb `URLSessionSwizzler` and 2 smart interceptors or 1 smart `Swizzler` and 2 dumb interceptors?

does that make sense?

File: Tests/DatadogTests/Datadog/Core/AutoInstrumentation/URLSessionSwizzlerTests.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (buranmert) [reply Tests/DatadogTests/Datadog/Core/AutoInstrumentation/URLSessionSwizzlerTests.swift]: new implementation of `dataTaskWithURL` works like that:
```
let task = currentImp(...)
if interceptor(task.request) {
  interceptTask(task)
}
return task
```
it's not concerned of headers or any other details of interception 👌

File: Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:161]: > if we follow [this way](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-452-propose-swizzling-refactoring/Tests/DatadogTests/Datadog/Core/Autoinstrumentation/URLSessionSwizzlerTests.swift) then we will need to test twice if `TracingInterceptor` and `RUMInterceptor` keep track of intercepted tasks properly.
here we implement and test this within `URLSessionSwizzler` once so that interceptors won't need to know `URLSession` internals.

No. Following the link, the `URLSessionSwizzlerTests` are tracing-agnostic. They use `URLSessionInterceptorMock` conforming to `URLSessionInterceptor` and test that the interceptor methods are properly called on the `URLSessionInterceptorMock`. It's a classic dependency-inversion making the swizzler depend on interface, not implementation.

By asserting that swizzler invokes the methods of `URLSessionInterceptor` interface in the right order, we can then build two interceptor objects:
* [`URLSessionTracingInterceptor: URLSessionInterceptor`](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-452-propose-swizzling-refactoring/Sources/Datadog/Tracing/Autoinstrumentation/URLSessionTracingInterceptor.swift),
*  `URLSessionRUMInterceptor: URLSessionInterceptor`

and test them individually to assert that both do the right instrumentation job.

### Cleariness

Both interceptors will have the same interface, **enforced by protocol**:
```swift
internal protocol URLSessionInterceptor {
    /// Tell if the interception should be run for given `URL`.
    func shouldIntercept(url: URL) -> Bool

    /// Tell if the interception should be run for given `URLRequest`.
    func shouldIntercept(request: URLRequest) -> Bool

    /// Return modified `URLRequest`
    func intercept(request: URLRequest) -> URLRequest

    /// Do anything when request's task is started.
    func notifyStart(of task: URLSessionTask)

    /// Return modified `URLResponse`
    func intercept(response: URLResponse?, of task: URLSessionTask?) -> URLResponse?
}
```

To create any new interceptor, you just take the protocol and conform to it. It's API is self-guiding.

To base interception on logic arithmetics, this can be scaled up to `ConditionalURLSessionInterceptor`, `CombinedURLSessionInterceptor` wrapping any group of interceptors.

To chain interceptors, `URLSessionInterceptorChain` can be introduced, also conforming to the same interface, but passing data through chained interceptor objects.

If any new interceptor requires a state, we can simply add it internally on the interceptor object.

Now, compare it to the existing hooks-approach:
```swift
internal typealias RequestInterceptor = (URLRequest) -> InterceptionResult?
internal struct InterceptionResult {
    let modifiedRequest: URLRequest
    let taskObserver: TaskObserver
}

internal typealias TaskObserver = (TaskObservationEvent) -> Void
internal enum TaskObservationEvent {
    case starting(URLRequest?)
    case completed(URLResponse?, Error?)
}
```
and think of how it scales up to any of mentioned extensions. It doesn't. Static functions are stateless, the `InterceptionResult` concept doesn't allow for grouping interceptor calls, any conditional logic of interception would have to be considered right in the implementation of a concrete interceptor. A vast refactoring would be required if only we need to go there.

"But we don't have that use cases" - you may say. Yes we don't but I'm proposing is not a fancy new approach which I authored. It's an obvious and popular way of handling network interception using interceptor pattern by passing `request` and `response` data back and forth ([iOS](https://github.com/Alamofire/Alamofire/blob/a57387a4b268304b14e04d51a462f4b570ee978a/Source/RequestInterceptor.swift#L155-L194), [Android](https://github.com/square/okhttp/blob/b086b8d1835bcb3825d60ceb44bec7dec356405b/docs/interceptors.md), [Web](https://angular.io/api/common/http/HttpInterceptor)). So instead of authoring our own, narrowly tailored implementation, why don't we follow what's widely used, what [others do at Datadog](https://github.com/DataDog/dd-sdk-android/blob/4b2234f97c01629ac6fbbf03def62be45638f150/dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptor.kt#L50-L88) and what we do in other parts of the codebase (protocols + dependency injection [1](https://github.com/DataDog/dd-sdk-ios/blob/master/Sources/Datadog/Logging/LogOutputs/LogFileOutput.swift) / [2](https://github.com/DataDog/dd-sdk-ios/blob/master/Sources/Datadog/Logging/LogOutputs/LogUtilityOutputs.swift) / [3](https://github.com/DataDog/dd-sdk-ios/blob/master/Sources/Datadog/Tracing/SpanOutputs/SpanFileOutput.swift))?

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:161]: first of all, the first version of PR had a wrong fix: i made a very quick fix and i did it in the wrong place. that could be confusing for you and sorry about that. you may wanna have another look to it before read the rest

there are too many wrongs here, i'll try to explain briefly but if you really think this PR is not a step in the right direction, please raise a PR and i will approve it real quick 🙏 

of course we can make and use interceptor chains
```swift
combinedInterceptor(interceptors: [RequestInterceptor]) -> RequestInterceptor {
  return { request in
    let modifiedRequest = interceptors.reduce(request) { $0.doSomething() }
    return modifiedRequest
  }
}
```
the interface is simple, 1 method: "i give you 1 request and you return me another request. either chain other interceptors or use a random request generator, i simply don't care"

we can change `typealias RequestInterceptor` to `protocol RequestInterceptor { func intercept(request) -> (Request, Observer)`, this is a trivial change and doesn't change anything IRL

`RequestInterceptor` block can have its own internal state with captured variables, but the aim is making it as pure as possible: `request ->  request`. it can be a class as well if that solves a problem

to summarize your comments in this PR:
1. don't do task-to-span mapping in swizzler, let interceptors do it
2. use 3 separate methods to intercept a request

i try to say:
1. instead of doing mapping twice (2 interceptors), let's do it once in swizzler
2. let's have one single method to reduce cyclomatic complexity in code

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:161]: you can also have a look at interceptor tests [here](https://github.com/DataDog/dd-sdk-ios/pull/139/files#diff-4349f34c8b14f81a69be1446a720897fR22)

[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:161]: > of course we can make and use interceptor chains
> ```swift
> combinedInterceptor(interceptors: [RequestInterceptor]) -> RequestInterceptor {
>  return { request in
>    let modifiedRequest = interceptors.reduce(request) { $0.doSomething() }
>    return modifiedRequest
>  }
> }
> ```
> the interface is simple, 1 method: "i give you 1 request and you return me another request. either chain other interceptors or use a random request generator, i simply don't care"

This only modifies the `request`, it's not possible to intercept the `response` as it's part of the `TaskObservationEvent`.

> we can change `typealias RequestInterceptor` to `protocol RequestInterceptor { func intercept(request) -> (Request, Observer)`, this is a trivial change and doesn't change anything IRL

This is exactly what I mean - to further extend this concept, you need to change the interface and all the existing implementations.

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Core/AutoInstrumentation/URLSessionSwizzler.swift:161]:
```swift
// pseudo-code
combinedObserver(_ observers: [Observer]) -> Observer {
  return { event in
     observers.forEach { $0.(event) }
  }
}
combinedInterceptor(interceptors: [RequestInterceptor]) -> RequestInterceptor {
 return { request in
   var observers = [Observer]()
   let modifiedRequest = interceptors.reduce(request) { $0.doSomething(); observers.append($0.observer); }
    let combinedObserver = combinedObserver(observers)
   return (modifiedRequest, combinedObserver)
 }
}
```
or that doesn't please you, although we don't have a use case for this in the foreseeable future, instead of adding new code we can simply change `InterceptionResult = (Request, [Observer])`
```swift
// TracingInterceptor
// before
return (modifiedRequest, observer)
// after
return (modifiedRequest, [observer])
...all the same old code...
```
that requires adding 2 chars per interceptors. if we have 1000 interceptors, that will require 2000 lines of code change.

i mean we DO NOT HAVE TO change it but we CAN change it. even if we change, it doesn't make any difference block, protocol, class, whatever

if there are other things that you think is not possible, can you please ask first in DM? 🙏 
i'll be happy to explain everything that you think is not possible in DM but here this PR becomes un-followable :(
